### PR TITLE
[analyzer] Improve bug report hashing, merge similar reports

### PR DESCRIPTION
**Pull request by NagyDonat**
Previously there were certain situations where alpha.security.ArrayBoundV2 produced lots of very similar and redundant reports that only differed in their full `Description` that contained the (negative) byte offset value. (See https://github.com/llvm/llvm-project/issues/86969 for details.)

This change updates the `Profile()` method of `PathSensitiveBugReport` to ensure that it uses `getShortDescription()` instead of the full `Description` so the standard report deduplication eliminates most of these redundant reports.

Note that the effects of this change are very limited because there are very few checkers that specify a separate short description, and so `getShortDescription()` practically always defaults to returning the full `Description`.

For the sake of consistency `BasicBugReport::Profile()` is also updated to use the short description. (If I recall correctly there are no checkers that use `BasicBugReport` with separate short and long descriptions.)

This commit also includes some small code quality improvements in `ArrayBoundV2` that are IMO too trivial to be moved into a separate commit.

### CODE REVIEW

File: clang/lib/StaticAnalyzer/Checkers/ArrayBoundCheckerV2.cpp [hunk 1]
```diff
@@ -609,7 +610,7 @@ void ArrayBoundCheckerV2::performCheck(const Expr *E, CheckerContext &C) const {
   // CHECK UPPER BOUND
   DefinedOrUnknownSVal Size = getDynamicExtent(State, Reg, SVB);
   if (auto KnownSize = Size.getAs<NonLoc>()) {
-    // In a situation where both overflow and overflow are possible (but the
+    // In a situation where both underflow and overflow are possible (but the
```
Comment (NagyDonat) [inline clang/lib/StaticAnalyzer/Checkers/ArrayBoundCheckerV2.cpp:613]: This is an unrelated change, I just noticed a mistake (that was introduced in one of my earlier commits).

**Review by steakhal — COMMENTED** [review]
This makes sense.
I wonder if we could have something in between. I'm thinking of having the concrete offset as a separate note, instead of having it part of the primary message.
That way after BR selection, we would still deterministically pick the shortest parh, and also have the offset in the path notes.

I find this better than dropping the concrete offset you worked so hard acquiring. WDYT?

**Comment by NagyDonat** [comment]
> I wonder if we could have something in between. I'm thinking of having the concrete offset as a separate note, instead of having it part of the primary message. That way after BR selection, we would still deterministically pick the shortest parh, and also have the offset in the path notes.

Actually I had a better solution for achieving this: during my earlier experimentation I noticed that the `Profile()` methods of the `BugReporter` use the (long) `Description` field (which is "Access of `<region>` at negative byte offset -124", the final note on the bug path), but ignore the `ShortDescription` field (which is "Out of bound access to memory preceding `<region>`", the stand-alone warning message). In the commit that I just pushed, I tweaked `Profile()` to ensure that it uses the short description and ignores the long one -- which yields the right deduplication behavior.

I'd guess that `ShortDescription` was missing from the `Profile()` method because it was introduced at a later point in the development, and there are very few checkers that actually set different short and long descriptions.

(For more details see the commit message of the commit that I just pushed onto this PR.)

**Comment by NagyDonat** [comment]
_(Technical detail: I'll be on vacation during the next week, so I won't see updates on this PR until the 22nd of July. If you want to merge this PR, feel free to do so, but don't forget to adjust the description and the title to accurately reflect the current state of the PR.)_

File: clang/lib/StaticAnalyzer/Checkers/ArrayBoundCheckerV2.cpp [hunk 2]
```diff
@@ -373,14 +373,14 @@ static std::optional<int64_t> getConcreteValue(std::optional<NonLoc> SV) {
 }
 
 static Messages getPrecedesMsgs(const SubRegion *Region, NonLoc Offset) {
-  std::string RegName = getRegionName(Region);
-  SmallString<128> Buf;
-  llvm::raw_svector_ostream Out(Buf);
-  Out << "Access of " << RegName << " at negative byte offset";
-  if (auto ConcreteIdx = Offset.getAs<nonloc::ConcreteInt>())
-    Out << ' ' << ConcreteIdx->getValue();
```
Comment (NagyDonat) [inline clang/lib/StaticAnalyzer/Checkers/ArrayBoundCheckerV2.cpp:381]: This `raw_svector_ostream` solution is very clunky, so I replaced it with a less ugly alternative when I re-added the code that prints the offset.

**Comment by steakhal** [comment]
> _(Technical detail: I'll be on vacation during the next week, so I won't see updates on this PR until the 22nd of July. If you want to merge this PR, feel free to do so.)_ 

No need to rush. Have a nice one!

**Comment by NagyDonat** [comment]
I'm back from the vacation, so I would like to restart and conclude this review process. @steakhal or anybody else please review when it's convenient for you.

**Comment by NagyDonat** [comment]
(Btw it seems that I accidentally added everyone as an "assignee" instead of a reviewer...)

File: clang/lib/StaticAnalyzer/Core/BugReporter.cpp [hunk 3]
```diff
@@ -2213,7 +2213,7 @@ void BasicBugReport::Profile(llvm::FoldingSetNodeID& hash) const {
 void PathSensitiveBugReport::Profile(llvm::FoldingSetNodeID &hash) const {
   hash.AddInteger(static_cast<int>(getKind()));
   hash.AddPointer(&BT);
-  hash.AddString(Description);
+  hash.AddString(getShortDescription());
```
Comment (steakhal) [inline clang/lib/StaticAnalyzer/Core/BugReporter.cpp:2216]: Is it intentional that you use a member function here instead of the member variable?

**Review by steakhal — APPROVED** [review]
LGTM. I only had one minor question inline.

File: clang/lib/StaticAnalyzer/Core/BugReporter.cpp (hunk 3 above)
Comment (NagyDonat) [reply clang/lib/StaticAnalyzer/Core/BugReporter.cpp:2216]: Yes, it's intentional, because the method call `getShortDescription()` is equivalent to `ShortDescripton.empty() ? Description : ShortDescription`. (The common case when the short and full descriptions are identical is represented internally by an empty `ShortDescription` and the shared value stored in `Description`.)
